rustls-symcrypt/src/signer.rs

```diff
@@ -1,22 +1,394 @@
+            PrivateKeyDer::Pkcs1(pkcs1) => {
+                // Get the DER-encoded private key blob
+                let private_key_blob = pkcs1.secret_pkcs1_der(); // returns &[u8] that is DER encoded
+
+                // Parse the DER-encoded RSA private key
+                let private_key = RsaPrivateKey::from_der(private_key_blob)
+                    .map_err(|_| Error::General("Failed to parse PKCS#1 DER".into()))?;
+
+                // Extract the components and set the key pair
+                RsaKey::set_key_pair(
+                    private_key.modulus.as_bytes(),
+                    private_key.public_exponent.as_bytes(),
+                    private_key.prime1.as_bytes(),
+                    private_key.prime2.as_bytes(),
+                    RsaKeyUsage::Sign,
+                )
+                .map_err(|_| Error::General("Failed to set RsaKey from PKCS#1".into()))?
+            }
+
+            PrivateKeyDer::Pkcs8(pkcs8) => {
+                let private_key_blob = pkcs8.secret_pkcs8_der(); // DER-encoded &[u8]
+
+                // Parse the DER-encoded private key
+                let private_key_info = PrivateKeyInfo::from_der(private_key_blob)
+                    .map_err(|_| Error::General("Failed to parse PKCS#8 DER".into()))?;
+                let private_key = RsaPrivateKey::from_der(&private_key_info.private_key)
+                    .map_err(|_| Error::General("Failed to parse PKCS#8 DER".into()))?;
+
+                // Extract the components and set the key pair
+                RsaKey::set_key_pair(
+                    private_key.modulus.as_bytes(),
+                    private_key.public_exponent.as_bytes(),
+                    private_key.prime1.as_bytes(),
+                    private_key.prime2.as_bytes(),
+                    RsaKeyUsage::Sign,
+                )
+                .map_err(|_| Error::General("Failed to set RsaKey from PKCS#8".into()))?
+            }
+            _ => {
+                return Err(Error::General(
+                    "Failed to parse RSA private key as either PKCS#1 or PKCS#8".into(),
+                ));
+            }
+        };
+
+        Ok(Self { key: Arc::new(key) })
+    }
+}
+
+impl SigningKey for RsaSigningKey {
+    fn choose_scheme(&self, offered: &[SignatureScheme]) -> Option<Box<dyn Signer>> {
+        ALL_RSA_SCHEMES
+            .iter()
+            .find(|scheme| offered.contains(scheme))
+            .map(|&scheme| {
+                Box::new(RsaSigner {
+                    key: Arc::clone(&self.key),
+                    scheme,
+                }) as Box<dyn Signer>
+            })
+    }
+
+    fn algorithm(&self) -> SignatureAlgorithm {
+        SignatureAlgorithm::RSA
+    }
+}
+
+#[derive(Debug)]
+struct RsaSigner {
+    key: Arc<RsaKey>,
+    scheme: SignatureScheme,
+}
+
+impl Signer for RsaSigner {
+    fn sign(&self, message: &[u8]) -> Result<Vec<u8>, Error> {
+        // from Ring, do we need to add padding???
```

Comment:
Did we resolve this padding issue? Also remove this comment

Reply:
done